cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1076 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "detail/graph_partition_utils.cuh"
+#include "prims/kv_store.cuh"
+#include "utilities/collect_comm.cuh"
+
+#include <cugraph/detail/collect_comm_wrapper.hpp>
+#include <cugraph/detail/decompress_edge_partition.cuh>
+#include <cugraph/detail/shuffle_wrappers.hpp>
+#include <cugraph/graph.hpp>
+#include <cugraph/src_dst_lookup_container.hpp>
+#include <cugraph/utilities/mask_utils.cuh>
+#include <cugraph/utilities/misc_utils.cuh>
+
+#include <raft/core/handle.hpp>
+
+namespace cugraph {
+
+template <typename edge_id_t, typename edge_type_t, typename value_t>
+template <typename _edge_id_t, typename _edge_type_t, typename _value_t>
+struct search_container_t<edge_id_t, edge_type_t, value_t>::search_container_impl {
+  static_assert(std::is_integral_v<edge_id_t>);
+  static_assert(std::is_integral_v<edge_type_t>);
+
+  static_assert(std::is_same_v<edge_type_t, _edge_type_t>);
+  static_assert(std::is_same_v<edge_id_t, _edge_id_t>);
+  static_assert(std::is_same_v<value_t, _value_t>);
+
+  ~search_container_impl() {}
+  search_container_impl() {}
+  search_container_impl(raft::handle_t const& handle,
+                        std::vector<edge_type_t> typs,
+                        std::vector<edge_id_t> typ_freqs)
+  {
+    auto invalid_key   = cugraph::invalid_vertex_id<edge_id_t>::value;
+    auto invalid_value = invalid_of_thrust_tuple_of_integral<value_t>();
+
+    edge_type_to_kv_store = container_t{};
+    edge_type_to_kv_store.reserve(typs.size());
+
+    for (size_t idx = 0; idx < typs.size(); idx++) {
+      auto typ              = typs[idx];
+      size_t store_capacity = typ_freqs[idx];
+
+      edge_type_to_kv_store.insert(
+        {typ, std::move(store_t(store_capacity, invalid_key, invalid_value, handle.get_stream()))});
+    }
+  }
+
+  void insert(raft::handle_t const& handle,
+              edge_type_t type,
+              raft::device_span<edge_id_t const> edge_ids_to_insert,
+              decltype(cugraph::allocate_dataframe_buffer<value_t>(
+                0, rmm::cuda_stream_view{}))&& values_to_insert)
+  {
+    auto itr = edge_type_to_kv_store.find(type);
+
+    if (itr != edge_type_to_kv_store.end()) {
+      assert(itr->first == type);
+
+      itr->second.insert(edge_ids_to_insert.begin(),
+                         edge_ids_to_insert.end(),
+                         cugraph::get_dataframe_buffer_begin(values_to_insert),
+                         handle.get_stream());
+
+    } else {
+      assert(false);
+    }
+  }
+
+  std::optional<decltype(cugraph::allocate_dataframe_buffer<value_t>(0, rmm::cuda_stream_view{}))>
```

Comment:
Why is this return type optional?  I don't see a case where it's never returned.

Reply:
Removed.